perun-wui-registrar/src/main/java/cz/metacentrum/perun/wui/registrar/widgets/items/validators/SshKeysListBoxValidator.java

```diff
@@ -26,34 +33,104 @@ public boolean validateLocal(ListBox listBox) {
 			listBox.setRawStatus(getTransl().cantBeEmpty(), ValidationState.ERROR);
 			return false;
 		}
+		// FIXME - but it is probably not necessary as API should check it
+		for (ExtendedTextBox extendedTextBox : listBox.getListValue()) {
+			String sshKey = extendedTextBox.getValue();
+			if (sshKey.contains(",")) {
+				setResult(Result.INVALID_FORMAT);
+				listBox.setStatus(getTransl().sshKeySeparatorNotAllowed(), ValidationState.ERROR);
+				return false;
+			}
+		}
+		listBox.setStatus(ValidationState.SUCCESS);
+		return true;
+	}
+
+	@Override
+	public void validate(ListBox listBox, Events<Boolean> events) {
+
+		if (!validateLocal(listBox)) {
+			events.onFinished(false);
+			return;
+		}
+
+		if (listBox.getValue() == null || listBox.getValue().isEmpty()) {
+			events.onFinished(true);
+			return;
+		}
+
+		checkVals.clear();
+		indexMap.clear();
+		callFailed = false;
 
-		if (listBox.getValue() != null && !listBox.getValue().isEmpty()) {
+		int counter = 1;
+		for (ExtendedTextBox extendedTextBox : listBox.getListValue()) {
+			String sshKey = extendedTextBox.getValue();
+			indexMap.put(sshKey, counter);
+			checkVals.put(sshKey, ValidationState.NONE);
+			counter++;
+		}
 
-			String wrongValues = "";
-			int index = 1;
-			for (ExtendedTextBox extendedTextBox : listBox.getListValue()) {
-				String sshKey = extendedTextBox.getValue();
+		for (String sshKey : checkVals.keySet()) {
 
-				if (sshKey.contains(",")) {
-					setResult(Result.INVALID_FORMAT);
-					listBox.setStatus(getTransl().sshKeySeparatorNotAllowed(), ValidationState.ERROR);
-					return false;
+			UsersManager.validateSSHKey(sshKey, new JsonEvents() {
+				@Override
+				public void onFinished(JavaScriptObject result) {
+					validatingCounter--;
+					checkVals.put(sshKey, ValidationState.SUCCESS);
+					if (validatingCounter == 0) {
+						// last check trigger events
+						for (ValidationState state : checkVals.values()) {
+							// at least one key is invalid -> switch to error if API call didn't fail
+							if (ValidationState.ERROR == state && !callFailed) {
```

Comment:
Check on `callFailed` should be used inside this if as another if used only to determine form item result and message. Otherwise item status is switched to success.
```java
if (ValidationState.ERROR == state) {
	if (!callFailed) {
		setResult(Result.INVALID_FORMAT);
		listBox.setRawStatus(getTransl().incorrectFormatItemList() + " <b> <br>" + checkVals.entrySet().stream().filter(entry -> (entry.getValue() == ValidationState.ERROR)).map((entry) -> indexMap.get(entry.getKey()) + ". " +
			(entry.getKey().length() > 25 ? entry.getKey().substring(0, 23) + "..." : entry.getKey())).collect(Collectors.joining("<br>")) + "</b>", ValidationState.ERROR);
	}
	// pass to outer event as fail
	events.onFinished(false);
	return;
}
```